vyper/venom/passes/loop_invariant_hosting.py

```diff
@@ -0,0 +1,127 @@
+    dfg: DFGAnalysis
+
+    def run_pass(self):
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        self.dfg = self.analyses_cache.request_analysis(DFGAnalysis)
+        loops = self.analyses_cache.request_analysis(LoopDetectionAnalysis)
+        self.loops = loops.loops
+        while True:
+            change = False
+            for from_bb, loop in self.loops.items():
+                hoistable: list[
+                    tuple[IRBasicBlock, IRBasicBlock, IRInstruction]
+                ] = self._get_hoistable_loop(from_bb, loop)
+                if len(hoistable) == 0:
+                    continue
+                change |= True
+                self._hoist(hoistable)
+            if not change:
+                break
+            # I have this inside the loop because I dont need to
+            # invalidate if you dont hoist anything
+            self.analyses_cache.invalidate_analysis(LivenessAnalysis)
+
+    def _hoist(self, hoistable: list[tuple[IRBasicBlock, IRBasicBlock, IRInstruction]]):
+        for loop_idx, bb, inst in hoistable:
+            bb.remove_instruction(inst)
+            bb_before: IRBasicBlock = loop_idx
+            bb_before.insert_instruction(inst, index=len(bb_before.instructions) - 1)
+
+    def _get_hoistable_loop(
+        self, from_bb: IRBasicBlock, loop: OrderedSet[IRBasicBlock]
+    ) -> list[tuple[IRBasicBlock, IRBasicBlock, IRInstruction]]:
+        result: list[tuple[IRBasicBlock, IRBasicBlock, IRInstruction]] = []
+        for bb in loop:
+            result.extend(self._get_hoistable_bb(bb, from_bb))
+        return result
+
+    def _get_hoistable_bb(
+        self, bb: IRBasicBlock, loop_idx: IRBasicBlock
+    ) -> list[tuple[IRBasicBlock, IRBasicBlock, IRInstruction]]:
+        result: list[tuple[IRBasicBlock, IRBasicBlock, IRInstruction]] = []
+        for instruction in bb.instructions:
+            if self._can_hoist_instruction(instruction, self.loops[loop_idx]):
+                result.append((loop_idx, bb, instruction))
+
+        return result
+
+    def _can_hoist_instruction(
+        self, instruction: IRInstruction, loop: OrderedSet[IRBasicBlock]
+    ) -> bool:
+        if _ignore_instruction(instruction):
+            return False
+        for bb in loop:
+            if self._in_bb(instruction, bb):
+                return False
+
+        if _is_correct_store(instruction):
+            for used_instruction in self.dfg.get_uses(instruction.output):
+                if not self._can_hoist_instruction_ignore_stores(used_instruction, loop):
+                    return False
+
+        return True
+
+    def _in_bb(self, instruction: IRInstruction, bb: IRBasicBlock):
+        for in_var in instruction.get_input_variables():
+            assert isinstance(in_var, IRVariable)
+            source_ins = self.dfg._dfg_outputs[in_var]
+            if source_ins in bb.instructions:
+                return True
+        return False
+
+    def _can_hoist_instruction_ignore_stores(
```

Comment:
leave a comment explaining why this is necessary?

Reply:
maybe just call this `_can_hoist_instruction` (or just `_can_hoist`!), as it doesn't need to be contrasted with a function like `_can_hoist_instruction_not_ignoring_stores`.